nex-rs/src/server/server_trait.rs

```diff
@@ -8,18 +8,14 @@ use async_trait::async_trait;
 use no_std_io::{StreamContainer, StreamWriter};
 use rand::RngCore;
 use std::{net::SocketAddr, sync::Arc, time::Duration};
-use tokio::{
-    net::UdpSocket,
-    sync::{Mutex, RwLock},
-    time,
-};
+use tokio::{net::UdpSocket, sync::RwLock, time};
 
 #[async_trait]
 pub trait Server: EventHandler {
     fn get_base(&self) -> &BaseServer;
     fn get_mut_base(&mut self) -> &mut BaseServer;
 
-    fn get_clients(&self) -> Arc<Mutex<Vec<ClientConnection>>> {
+    fn get_clients(&self) -> Arc<RwLock<Vec<Arc<RwLock<ClientConnection>>>>> {
```

Comment:
Why do we need to wrap the client connection locks with `Arc`?

Reply:
We can't pass a mutable reference of a client connection from the vector unless we have a write lock on it. We'd be changing the actual contents of the vector. By wrapping it in an RwLock, we are saying to hold a vector of references rather than actual objects. From there we can get read or write locks on that actual object using that reference.

Because we are only holding a write lock on the ClientConnection and clients can only operate one packet at a time anyway, this could also be changed to a mutex because we don't need to guarantee sync for the object.

Reply:
I understand you question a bit more now. Still goes the same as we can't pass references to an RwLock as it is owned by the vector

Reply:
References allow to use use data without ownership:
```rs
let addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 8080);
let client = ClientConnection::new(addr, ClientContext::default());

// Immutable RwLock clients list
let clients_lock: RwLock<Vec<RwLock<ClientConnection>>> = RwLock::new(vec![RwLock::new(client)]);

// Immutable Read guard
let clients = clients_lock.read().await;

// Mutable reference to client
let client = &mut clients[0].write().await;
client.set_is_connected(true);
```

Arcs allow us to have ownership in cases where we normally couldn't, but it doesn't look like we need that for the individual clients.